### PR TITLE
DOPE-226: added array functions, CM extension functions and tests

### PR DESCRIPTION
**Pull request by martinagallati-ergon**
**Info for ARRAY_PREPEND**: In the documentation for the function, the values precede the array expression. To implement the variable number of arguments I used vararg, which would impose a need to use a named argument for `array`. For usability's sake I did it analogous to `ARRAY_APPEND(array, vararg values)`, so a DOPE user does not need to name the array argument.

### CODE REVIEW

File: core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt [hunk 1]
```diff
@@ -1,6 +1,6 @@
 package ch.ergon.dope.validtype
 
-interface ValidType
+sealed interface ValidType
```
Comment (jansigi) [inline core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt:3]: Why is it sealed?

**Review by jansigi — APPROVED** [review]

File: core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt (hunk 1 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt:3]: On request of @m0ne, I changed it to sealed.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayConcatExpression.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package ch.ergon.dope.resolvable.expression.unaliased.type.arrayfunction
+
+import ch.ergon.dope.resolvable.expression.TypeExpression
+import ch.ergon.dope.validtype.ArrayType
+import ch.ergon.dope.validtype.ValidType
+
+class ArrayConcatExpression<T : ValidType>(
+    firstArray: TypeExpression<ArrayType<T>>,
+    secondArray: TypeExpression<ArrayType<T>>,
+    vararg arrays: TypeExpression<ArrayType<T>>,
```
Comment (m0ne) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayConcatExpression.kt:10]: Not the biggest fan of the naming here, maybe something like `additionalArrays` or something like that.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayContainsExpression.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import ch.ergon.dope.validtype.ArrayType
+import ch.ergon.dope.validtype.BooleanType
+import ch.ergon.dope.validtype.NumberType
+import ch.ergon.dope.validtype.StringType
+import ch.ergon.dope.validtype.ValidType
+
+class ArrayContainsExpression<T : ValidType>(array: TypeExpression<ArrayType<T>>, value: TypeExpression<T>) :
+    ArrayFunctionNumberExpression<T>("ARRAY_CONTAINS", array, value)
+
+fun <T : ValidType> arrayContains(array: TypeExpression<ArrayType<T>>, value: TypeExpression<T>) = ArrayContainsExpression(array, value)
```
Comment (m0ne) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayContainsExpression.kt:14]: Line on the verge of being too long, would split.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayContainsExpression.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+fun arrayContains(array: TypeExpression<ArrayType<StringType>>, value: String) = arrayContains(array, value.toDopeType())
+
+fun arrayContains(array: TypeExpression<ArrayType<NumberType>>, value: Number) = arrayContains(array, value.toDopeType())
+
+fun arrayContains(array: TypeExpression<ArrayType<BooleanType>>, value: Boolean) = arrayContains(array, value.toDopeType())
```
Comment (m0ne) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayContainsExpression.kt:20]: Are we leaving the new line between these functions now? What has been decided?

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        this.args = args
+        this.extra = extra
+    }
+
+    constructor(symbol: String, array: TypeExpression<ArrayType<T>>, vararg args: TypeExpression<ArrayType<out ValidType>>) {
+        this.symbol = symbol
+        this.array = array
+        this.args = args
+        extra = null
+    }
```
Comment (m0ne) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt:32]: Are you sure about the way you are doing a secondary constructor here? 
I would split the second constructor into multiple lines anyway, looks cleaner.

File: core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt (hunk 1 above)
Comment (m0ne) [reply core/src/main/kotlin/ch/ergon/dope/validtype/ValidType.kt:3]: Why shouldn't it be sealed?

**Review by m0ne — CHANGES_REQUESTED** [review]
In general, there are a few files that exceed the optimal length, none that exceed the maximum, but I would advise against longer lines than necessary. I would go through the PR again and break up long lines. Also, I noticed that the spacing between functions is now mixed (sometimes there is a blank line, sometimes there is none), check that again. 

And naming wise, there are places where we use     

firstArray: TypeExpression<ArrayType<T>>,
secondArray: TypeExpression<ArrayType<T>>,
vararg arrays: TypeExpression<ArrayType<T>>,

the name `arrays` is quite vague and it does not seem clear that we mean by arrays firstArray, secondArray, **thirdArray, fourthArray,...**

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayAppendExpression.kt [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+import ch.ergon.dope.validtype.ArrayType
+import ch.ergon.dope.validtype.ValidType
+
+class ArrayAppendExpression<T : ValidType>(
+    array: TypeExpression<ArrayType<T>>,
+    value: TypeExpression<T>,
+    vararg values: TypeExpression<T>,
+) : ArrayFunctionExpression<T>("ARRAY_APPEND", array, value, *values)
+
+fun <T : ValidType> arrayAppend(array: TypeExpression<ArrayType<T>>, value: TypeExpression<T>, vararg values: TypeExpression<T>) =
```
[the remark this answers is not in the document]
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayAppendExpression.kt:13]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayContainsExpression.kt (hunk 3 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayContainsExpression.kt:14]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayContainsExpression.kt (hunk 4 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayContainsExpression.kt:20]: I unified it to a space between functions.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayCountExpression.kt
[the remark this answers is not in the document]
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayCountExpression.kt:1]: I created a new ticket to evaluate the package structure and decide from there.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayCountExpression.kt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package ch.ergon.dope.resolvable.expression.unaliased.type.arrayfunction
+
+import ch.ergon.dope.resolvable.expression.TypeExpression
+import ch.ergon.dope.validtype.ArrayType
+import ch.ergon.dope.validtype.ValidType
+
+class ArrayCountExpression<T : ValidType>(array: TypeExpression<ArrayType<T>>) : ArrayFunctionNumberExpression<T>("ARRAY_COUNT", array)
```
[the remark this answers is not in the document]
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayCountExpression.kt:7]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt (hunk 5 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt:32]: It's not even needed anymore with my current implementation, so I removed it and made a primary constructor.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayConcatExpression.kt (hunk 2 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayConcatExpression.kt:10]: done

**Review by jansigi — APPROVED** [review]

**Review by m0ne — APPROVED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayAppendExpression.kt [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+class ArrayAppendExpression<T : ValidType>(
+    array: TypeExpression<ArrayType<T>>,
+    value: TypeExpression<T>,
+    vararg values: TypeExpression<T>,
+) : ArrayFunctionExpression<T>("ARRAY_APPEND", array, value, *values)
+
+fun <T : ValidType> arrayAppend(
+    array: TypeExpression<ArrayType<T>>,
+    value: TypeExpression<T>,
+    vararg values: TypeExpression<T>,
```
Comment (m0ne) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayAppendExpression.kt:16]: I think the same could be said here with Value, Values. Maybe a similar naming scheme you used before would make sense here: `value', `additionalValues', or something along those lines.

Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayAppendExpression.kt:16]: done, thanks 😃

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import ch.ergon.dope.DopeQuery
+import ch.ergon.dope.resolvable.expression.TypeExpression
+import ch.ergon.dope.resolvable.operator.FunctionOperator
+import ch.ergon.dope.validtype.ArrayType
+import ch.ergon.dope.validtype.ValidType
+
+sealed class ArrayFunctionExpression<T : ValidType>(
+    private val symbol: String,
+    private val array: TypeExpression<ArrayType<T>>,
+    private vararg val args: TypeExpression<out ValidType>,
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt:12]: don't use abbreviation for variable names, use _arguments_ here

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    private val extra: TypeExpression<out ValidType>? = null,
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt:13]: this 'extra' is only used in the _ArrayReplaceExpression_ (as far as I see). Can't you remove it already there, like e.g.:
```
class ArrayReplaceExpression<T : ValidType>(
    array: TypeExpression<ArrayType<T>>,
    toReplace: TypeExpression<T>,
    replaceWith: TypeExpression<T>,
    max: TypeExpression<NumberType>? = null,
) : ArrayFunctionExpression<T>("ARRAY_REPLACE", array, *listOfNotNull(toReplace, replaceWith, max).toTypedArray())
```

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        return DopeQuery(
+            queryString = toFunctionQueryString(symbol, arrayDopeQuery, *argsDopeQuery.toTypedArray(), extra = extraDopeQuery),
+            parameters = arrayDopeQuery.parameters + argsDopeQuery.fold(
+                emptyMap(),
+            ) { argsParameters, field -> argsParameters + field.parameters } + extraDopeQuery?.parameters.orEmpty(),
+        )
+    }
+}
+
+sealed class ArrayFunctionNumberExpression<T : ValidType>(
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt:28]: Why ArrayFunction**Number**Expression? Shouldn't it then be of type _TypeExpression< NumberType >_?

Comment (pgruntz) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt:28]: Maybe add some more expresive tests, where you do something like `...WHERE ARRAY_COUNT(...) + 7 < 20` - to check that we really have a number after calling such a function

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayPrependExpression.kt [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+import ch.ergon.dope.resolvable.expression.TypeExpression
+import ch.ergon.dope.resolvable.operator.FunctionOperator
+import ch.ergon.dope.validtype.ArrayType
+import ch.ergon.dope.validtype.ValidType
+
+class ArrayPrependExpression<T : ValidType>(
+    private val array: TypeExpression<ArrayType<T>>,
+    private val value: TypeExpression<T>,
+    private vararg val additionalValues: TypeExpression<T>,
+) : TypeExpression<ArrayType<T>>, FunctionOperator {
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayPrependExpression.kt:13]: Why cant you inherit from the _ArrayFunctionExpression_ here?

Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayPrependExpression.kt:13]: As written in the description, PREPEND behaves syntactically different from APPEND, so this approach made the most sense to me in this case (and after syncing with @m0ne about it).

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt (hunk 10 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayFunctionExpression.kt:13]: done

**Pull request merged**